render/vulkan/vulkan.c

```diff
@@ -0,0 +1,596 @@
+	uint32_t avail_extc = 0;
+	res = vkEnumerateDeviceExtensionProperties(phdev, NULL,
+		&avail_extc, NULL);
+	if (res != VK_SUCCESS || avail_extc == 0) {
+		wlr_vk_error("Could not enumerate device extensions (1)", res);
+		return NULL;
+	}
+
+	VkExtensionProperties avail_ext_props[avail_extc + 1];
+	res = vkEnumerateDeviceExtensionProperties(phdev, NULL,
+		&avail_extc, avail_ext_props);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("Could not enumerate device extensions (2)", res);
+		return NULL;
+	}
+
+	for (size_t j = 0; j < avail_extc; ++j) {
+		wlr_log(WLR_DEBUG, "Vulkan device extension %s v%"PRIu32,
+			avail_ext_props[j].extensionName, avail_ext_props[j].specVersion);
+	}
+
+	// create device
+	struct wlr_vk_device *dev = calloc(1, sizeof(*dev));
+	if (!dev) {
+		wlr_log_errno(WLR_ERROR, "allocation failed");
+		return NULL;
+	}
+
+	dev->phdev = phdev;
+	dev->instance = ini;
+	dev->drm_fd = -1;
+	dev->extensions = calloc(16 + ext_count, sizeof(*ini->extensions));
+	if (!dev->extensions) {
+		wlr_log_errno(WLR_ERROR, "allocation failed");
+		goto error;
+	}
+
+	// find extensions
+	for (unsigned i = 0; i < ext_count; ++i) {
+		if (find_extensions(avail_ext_props, avail_extc, &exts[i], 1)) {
+			wlr_log(WLR_DEBUG, "vulkan device extension %s not found",
+				exts[i]);
+			continue;
+		}
+
+		dev->extensions[dev->extension_count++] = exts[i];
+	}
+
+	// for dmabuf importing we require at least the
+	// 'external_memory_fd, external_memory_dma_buf, queue_family_foreign'
+	// extensions. So only enable them if all three are available (assumption
+	// throughout the codebase).
+	const char *names[] = {
+		VK_KHR_EXTERNAL_MEMORY_FD_EXTENSION_NAME,
+		VK_KHR_IMAGE_FORMAT_LIST_EXTENSION_NAME, // or vulkan 1.2
+		VK_EXT_EXTERNAL_MEMORY_DMA_BUF_EXTENSION_NAME,
+		VK_EXT_QUEUE_FAMILY_FOREIGN_EXTENSION_NAME,
+		VK_EXT_IMAGE_DRM_FORMAT_MODIFIER_EXTENSION_NAME,
+	};
+
+	unsigned nc = sizeof(names) / sizeof(names[0]);
+	const char *not_found = find_extensions(avail_ext_props, avail_extc, names, nc);
+	if (not_found) {
+		wlr_log(WLR_ERROR, "vulkan: required device extension %s not found",
+			not_found);
+		goto error;
+	}
+
+	for (unsigned i = 0u; i < nc; ++i) {
+		dev->extensions[dev->extension_count++] = names[i];
+	}
+
+	// queue families
+	{
+		uint32_t qfam_count;
+		vkGetPhysicalDeviceQueueFamilyProperties(phdev, &qfam_count, NULL);
+		assert(qfam_count > 0);
+		VkQueueFamilyProperties queue_props[qfam_count];
+		vkGetPhysicalDeviceQueueFamilyProperties(phdev, &qfam_count,
+			queue_props);
+
+		bool graphics_found = false;
+		for (unsigned i = 0u; i < qfam_count; ++i) {
+			graphics_found = queue_props[i].queueFlags & VK_QUEUE_GRAPHICS_BIT;
+			if (graphics_found) {
+				dev->queue_family = i;
+				break;
+			}
+		}
+
+		assert(graphics_found);
+	}
+
+	const float prio = 1.f;
+	VkDeviceQueueCreateInfo qinfo = {};
+	qinfo.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+	qinfo.queueFamilyIndex = dev->queue_family;
+	qinfo.queueCount = 1;
```

Comment:
This is not efficient. If you submit everything to a single queue when the compute family can have more, there's essentially no parallelism, and you may as well be using OpenGL.
You want to initialize all queues the queue family has, and rotate between them. That way, you can loosen up, and instead of waiting after each queue operation finishes, you can establish dependencies between operations (e.g. scaling each sub-window prior to compositing) via fences and semaphores.
Semaphores are optional, but really consider using them, since they're much faster and lower-latency than fences, and the way they're designed restricts you from using hacks, and if you make a mistake, it will deadlock. They're optional on paper, but really, all implementations support them, and both libplacebo and FFmpeg depend on them.

Reply:
> If you submit everything to a single queue when the compute family can have more, there's essentially no parallelism, and you may as well be using OpenGL.

This is definitely not true. As long as you don't have a serial data dependency, the GPU driver *can* schedule stuff even from a single queue. (Using VkEvents and appropriate pipeline barriers can help here)

In practice I have never measured a performance gain from using more than one queue.